submission/distributor_test.go

```diff
@@ -222,3 +271,126 @@ func TestNewDistributorRootPools(t *testing.T) {
 		})
 	}
 }
+
+func pemFileToDERChain(t *testing.T, filename string) [][]byte {
+	t.Helper()
+	rawChain, err := x509util.ReadPossiblePEMFile(filename, "CERTIFICATE")
+	if err != nil {
+		t.Fatalf("failed to load testdata: %v", err)
+	}
+	return rawChain
+}
+
+func getSCTMap(l []*AssignedSCT) map[string]*AssignedSCT {
+	m := map[string]*AssignedSCT{}
+	for _, asct := range l {
+		m[asct.LogURL] = asct
+	}
+	return m
+}
+
+// Stub CT policy to run tests.
+type stubCTPolicy struct {
+	baseNum int
+}
+
+// Builds simplistic policy requiring n SCTs from any Logs for each cert.
+func buildStubCTPolicy(n int) stubCTPolicy {
+	return stubCTPolicy{baseNum: n}
+}
+
+func (stubP stubCTPolicy) LogsByGroup(cert *x509.Certificate, approved *loglist.LogList) (ctpolicy.LogPolicyData, error) {
+	baseGroup, err := ctpolicy.BaseGroupFor(approved, stubP.baseNum)
+	groups := ctpolicy.LogPolicyData{baseGroup.Name: &baseGroup}
+	return groups, err
+}
+
+func TestDistributorAddPreChain(t *testing.T) {
+	testCases := []struct {
+		name     string
+		ll       *loglist.LogList
+		plc      ctpolicy.CTPolicy
+		rawChain [][]byte
+		scts     []*AssignedSCT
+		wantErr  bool
+	}{
+		{
+			name:     "MalformedChainRequest",
+			ll:       sampleValidLogList(t),
+			plc:      ctpolicy.ChromeCTPolicy{},
+			rawChain: pemFileToDERChain(t, "../trillian/testdata/subleaf.misordered.chain"),
+			scts:     nil,
+			wantErr:  true,
+		},
+		{
+			name:     "CallBeforeInit",
+			ll:       sampleValidLogList(t),
+			plc:      ctpolicy.ChromeCTPolicy{},
+			rawChain: nil,
+			scts:     nil,
+			wantErr:  true,
+		},
+		{
+			name:     "FullChainApplePolicy",
+			ll:       sampleValidLogList(t),
+			plc:      ctpolicy.AppleCTPolicy{},
+			rawChain: pemFileToDERChain(t, "../trillian/testdata/subleaf.chain"), // subleaf chain is fake-ca-1-rooted
+			scts: []*AssignedSCT{
+				{
+					LogURL: "ct.googleapis.com/rocketeer/",
+					SCT:    testSCT("ct.googleapis.com/rocketeer/"),
+				},
+			},
+			wantErr: true, // Not enough SCTs for policy
+		},
+		{
+			name:     "FullChain1Policy",
+			ll:       sampleValidLogList(t),
+			plc:      buildStubCTPolicy(1),
+			rawChain: pemFileToDERChain(t, "../trillian/testdata/subleaf.chain"),
+			scts: []*AssignedSCT{
+				{
+					LogURL: "ct.googleapis.com/rocketeer/",
+					SCT:    testSCT("ct.googleapis.com/rocketeer/"),
+				},
+			},
+			wantErr: false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			dist, _ := NewDistributor(tc.ll, tc.plc, buildStubLogClient)
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			defer cancel()
+
+			go dist.Run(ctx)
+			// First Log refresh expected.
+			<-ctx.Done()
+
+			scts, err := dist.AddPreChain(context.Background(), tc.rawChain)
+			if gotErr := (err != nil); gotErr != tc.wantErr {
+				t.Errorf("Expected to get errors is %v while actually getting errors is %v", tc.wantErr, gotErr)
```

Comment:
The conventional way to write a test error message for this would be as follows. If an error occurs unexpectedly, then the test output will contain the error message.
```suggestion
				t.Errorf("dist.AddPreChain() = (_, %v), want err? %t", err, tc.wantErr)
```

Reply:
Applied.

Reply:
It doesn't look like you applied this change consistently. There's a number of test error messages that still don't use this format.